### ISSUE TITLE
Provide a WebGPU build that re-exports from three

### ISSUE THREAD

**Issue by isaac-mason** [mrdoob/three.js#29156]
### Description

When using the WebGPU build introduced in r167 with [vite, using recommended aliases configuration](https://github.com/mrdoob/three.js/pull/28650#issuecomment-2198568721), I am unable to use renderer agnostic exports from libraries that also have webgl imports. Bundlers will have errors attempting to import webgl specific imports that are not exported by the webgpu build.

### Solution

If the three/webgpu build re-exported from three, aliases would not be required to use the webgpu build with bundlers, and libraries could export both webgl and webgpu utilities that import from both three and three/webgpu.

### Alternatives

A build that has both webgl and webgpu exports could solve some issues, but this is not as desirable as the proposed solution as bundler aliases would still be required to prevent duplication issues.

### Additional context

https://twitter.com/onirenaud/status/1824688713542341119

**Comment by hybridherbst** [comment]
+1 from me. 

Besides the missing exports, I think what also should be clarified is how libraries are supposed to start adding support for WebGPU – not everyone will be able/willing to provide/maintain separate bundles for webgl vs. webgpu.

That unclarity includes three.js examples themself – an example:
- GLTFExporter currently imports decompress, which is used to bring textures back from the GPU to the CPU
- `decompress` (in [TextureUtils](https://github.com/mrdoob/three.js/blob/dev/examples/jsm/utils/TextureUtils.js#L18)) imports WebGLRenderer
- is there a current plan for how that will work in WebGPURenderer, and how GLTFExporter will support both? Or will there be a hard cut at some point in time X, where GLTFExporter will switch over to WebGPURenderer?

**Comment by GitHubDragonFly** [comment]
+1 from me as well but there is a [forum topic](https://discourse.threejs.org/t/three-webgpu-min-js-to-provide-webglrenderer-export/69075) which suggests that this might not be happening.

**Comment by donmccurdy** [comment]
> is there a current plan for how that will work in WebGPURenderer, and how GLTFExporter will support both?

My feeling here would be that we may need to work out a way for GLTFExporter not to import a renderer. Either to require pre-processing the scene, or specifying some configuration on the exporter that provides a decompress implementation or renderer.

**Comment by Mugen87** [comment]
How about doing something like this:
```js
const exporter = new THREE.GLTFExporter();
exporter.setTextureUtils( utils ); 
```
`decompress()` also relies on `ShaderMaterial`. According to the previous migration, it's best to create a separate `TextureUtilsGPU` module and implement all helpers based on `WebGPURenderer` and node material.

`USDZExporter` requires the same change as well, btw.

**Comment by isaac-mason** [comment]
To clarify the proposal in this issue, regarding the forums response:

> The idea of separate builds is to avoid having both renderers in one file. So I would say no, it is not possible to have WebGLRenderer included in three.webgpu.min.js.
When WebGPURenderer more matures, there will be no need for WebGLRenderer anymore so you can rely on the built-in WebGL fallback.

This issue is not in opposition to having a WebGPU renderer build that does not contain WebGL renderer exports, this would primarily support use cases where bundlers are not used.

However, using that same build with bundler use cases is problematic, which is why this issue is asking for the `three/webgpu` entrypoint to re-export from `three`.

Something we can anticipate will be very common during the transition period is libraries exporting both WebGL and WebGPU utilities. Without making the proposed change, libraries cannot easily do this.

**Comment by Mugen87** [comment]
> If the three/webgpu build re-exported from three,

How does this work? Does the project need a specific configuration in its `package.json`?

**Comment by isaac-mason** [comment]
I imagine this could be approached like this:

1. WebGPU related code would need to import from `three`, not from `src`

2. The `Three.WebGPU.js` barrel file would change from exporting from `src`, to re-exporting from three, e.g. `export { ... } from 'three'` and exporting WebGPURenderer specific exports.

3. The `three/webgpu` and `three/tsl` entrypoints would change from `build/three.webgpu.js` (which only contains WebGPURenderer code) to the new `Three.WebGPU.js` barrel file, which imports and re-exports from `three`. At this point imports to e.g. `Mesh` from both `three` and `three/webgpu` would now resolve to the same class.o

4. The existing rollup build for `three.webgpu.js` that doesn't contain WebGLRenderer code could remain, but entrypoints in `package.json` wouldn't use it. Using the webgpu-only build in non-bundled projects wouldn't change, importmaps would still be used as they are now.

**Comment by mrdoob** [comment]
How about we introduce a new package?

Package `three`: WebGLRenderer and WebGLRenderer-Supported Addons
Package `three-next`: WebGPURenderer and WebGPURenderer-Supported Addons

We'd have to figure out how to publish two packages from a single repo...

Then, maybe in 10 years (😶)... We could make both the same?

/cc @sunag

**Comment by hybridherbst** [comment]
I think that would force each thing in the ecosystem to also provide multiple packages, instead of being able to support "three" with WebGL and WebGPU with one ecosystem package...

My opinion is that re-exporting with WebGPU, making sure the code in three supports both systems, and providing examples for how third-party code supports both systems would be better.

**Comment by donmccurdy** [comment]
Yes, I am worried `three-next` would break a great many things, and then require another large break to consolidate later on.

If we had _no_ other solution, then publishing two versions (not two packages) under `three` and `three@next` instead might be an option. A package that supports both can specify that in the peer dependencies ranges. But there are still a lot of implications to think through here, I don't see how a library could support both WebGLRenderer and WebGPURenderer at the same time, so it'd work better for something like three-pathtracing which only needs things like Object3D, BufferGeometry, and math.

The suggestion from @isaac-mason sounds workable to me, @mrdoob do you have 
particular concerns with that? It's mainly about making sure that WebGPURenderer can be used without a bundler, and without getting WebGL stuff by accident?

**Comment by CodyJasonBennett** [comment]
Another idea is re-exporting from `three.module.js` instead of `three` since we know WebGPU code relies on top-level-await and is therefore ESM only. Ideally, this doesn't change how tools will resolve or bundle three and should work OOTB for import maps. I've been keeping an eye on tree-shaking in core so WebGL stuff isn't pulled in by accident for this exactly #28670.

**Comment by donmccurdy** [comment]
Possibly related... do we need/want the `three/webgpu` and `three/tsl` entrypoints to be separate and identical? Unless there's a longer-term plan for that, I think we should pick one.

**Comment by seancheno** [comment]
I stumbled upon this post after upgrading our three version to 0.169 from 0.167 and switching around our exports along with our ts config paths and vite config alias.  We currently are using the WebGPURenderer entirely throughout our project.


vite config:

```
        alias: {
          'three/examples/jsm': 'three/examples/jsm',
          'three/addons': 'three/examples/jsm',
          'three/tsl': 'three/webgpu',
          'three': 'three/src/three.webgpu.js',
          },
```
          
  ts config:
  
```
      resolve: {
        alias: {
          'three/examples/jsm': 'three/examples/jsm',
          'three/addons': 'three/examples/jsm',
          'three/tsl': 'three/webgpu',
          'three': 'three/src/three.webgpu.js',
          },
      },
```
      
 I wasn't sure if we were doing this entirely correct, but everything seemed to work fine. By doing this along with changing a few import statements, it also allowed us to get rid of the "multiple instances of webgpu being imported" that we had originally. 
 
 The only issue we face at the moment is when importing and using the GLTFExporter:
 
` import { GLTFExporter } from "three/examples/jsm/exporters/GLTFExporter";
`
Since the GLTFExporter uses the decompress method of TextureUtils, we get the error 

node_modules/three/examples/jsm/utils/TextureUtils.js (10:1): "WebGLRenderer" is not exported by "node_modules/three/src/Three.WebGPU.js", imported by "node_modules/three/examples/jsm/utils/TextureUtils.js".

Which stems from this line in the decompress method:

```
if ( renderer === null ) {

     renderer = _renderer = new WebGLRenderer( { antialias: false } );

}
```
	
If we comment out that line, it removes the error and we are then able to use the GLTFExporter without any issues.

I see that there is the TextureUtilsGPU.js file now, but that it's not used anywhere within the three code. I apologize if I'm missing something obvious here, but wanted to bump this thread to showcase how we are unable to use the GLTFExporter at the moment unless we comment out that line. For the time being we will be creating a mod of the GLTFExporter and commenting out that line.

Thanks for taking another look into this.

**Comment by donmccurdy** [comment]
Hoping to fix GLTFExporter with:

- https://github.com/mrdoob/three.js/pull/29509

But there are more issues still to resolve in that PR.

And, I think we do need to agree on https://github.com/mrdoob/three.js/pull/29404 or something along those lines, so that custom vite configs are not required.

**Comment by RenaudRohlinger** [comment]
Bumping this as a solution like this is crucial for us to move forward with the WebGPURenderer.

The approach I suggested in [this comment](https://github.com/mrdoob/three.js/pull/28650#issuecomment-2198568721) is more of a temporary patch that sidesteps the core issue. A solid implementation is needed to ensure the WebGPURenderer integrates seamlessly with the Three.js ecosystem and its libraries.

Could we consider advancing with this? /cc @sunag @mrdoob @Mugen87

**Comment by Mugen87** [comment]
I'm okay with https://github.com/mrdoob/three.js/pull/29404. If the merge conflicts are resolved, we can make a merge from my point of view.

**Comment by LukeWood** [comment]
Not sure if this is still a point of discussion - but I was just considering how much nicer it would be to have a completely distinct package for the gpu stuff.  It's a huge head-ache trying to make sure you don't accidentally import `three` when you want `three/webgpu`, and I think the best thing we could do is "stop pretending" i.e. just have a fully distinct package that only exports webgpu compatible things.

Obviously, this might be super hard to implement in practice - but figured I'd call out that I was just wishing for this (which is how I found this issue)

**Comment by isaac-mason** [comment]
> It's a huge head-ache trying to make sure you don't accidentally import three when you want three/webgpu

#29404 will address that head-ache by making it completely fine to import renderer agnostic imports from `three` when using `three/webgpu`. e.g. `Vector3` from `three` and `three/webgpu` will have the same identity.

**Comment by LukeWood** [comment]
Oh fantastic - ok then problem solved from my perspective.  Vector, Box, Frustrum and Sphere have been the sources of my woes in this area.

**Comment by trusktr** [comment]
> However, using that same build for bundler use cases is problematic, which is why this issue is asking for the `three/webgpu` entrypoint to re-export from `three`.

Simpler solution:

- `three` exports everything, easy to use, people can select `WebGLRenderer` or `WebGPURenderer`. Both of them will exist for quite a long time
- make two new paths for people that want them separate who do not have tree shaking build tools: `three/webgl` and `three/webgpu`

This approach keeps things working very easily for anyone already using `three` (most people). People who want a slimmer library who do not have a tree shaking build can pick one of the other two. People who know how to configure build tool aliases can point `three` to either `three/webgl` or `three/webgpu`.

But the best thing about this approach is that the huge amounts of things already build on `three` will have a very simple way to start trying WebGPU **without modifying any build systems** by changing this,

```js
import {WebGLRenderer} from 'three'
```

to this,

```js
import {WebGPURenderer} from 'three'
```

or even to this if they want to use both (f.e. libraries that provide multiple render modes):

```js
import {WebGPURenderer, WebGPURenderer} from 'three'
```

---

I know a big company (one of the giants you've all heard of) has a build step that ingests the `three` lib, and now it would need to be modified to be able to support WebGPU, which is a big pain that doesn't need to exist.

**Comment by trusktr** [comment]
@CodyJasonBennett The solution in https://github.com/mrdoob/three.js/pull/29404 still requires some people to update their custom builds of threejs (they need to provide new outputs in order to include webgpu). What do you think about the approach of the previous comment? It would allow all build setups to continue working as-is. I already know that the solution in #29404 will still require updates to build setups for existing projects I have access to.

**Comment by CodyJasonBennett** [comment]
> @CodyJasonBennett The solution in https://github.com/mrdoob/three.js/pull/29404 still requires people to update their builds (f.e. at some-big-company-you-know, this is a pain right now). Some of those builds produce outputs for internal org use, so now they are forced to figure out how to handle the webgpu path, when really the whole issue could be 100% avoided without build changes anywhere.

That is mistaken. #29404 is one of a few ways that require no build configuration. I suspect the issue is exactly they were relying on `three/webgpu` and `three` being the same, or more specifically, their configuration. I know this company, SpaceX, but I'd really appreciate it if they stopped trying to extort me and my colleagues around WebGPU and actually contribute or hold a conversation in an open forum. It shouldn't have to be said, but nothing is going to happen otherwise. It might be news to them that WebGPU is not a living standard, and all code you write can and will break until then. Unless you ship an implementation like wgpu/Dawn, it is extremely irresponsible to use in production, despite the WebGL fallback.

> Another thing is, we probably all expect the three export will eventually export the WebGPU stuff. The suggested approach aligns with that. People using the new APIs will not have to change their build setups again in the future to go back to three from three/webgpu. One day you might even want to delete three/webgpu and force another build setup update, for which there is no need.

Merging everything to `three` is an extremely breaking change that the ecosystem is not yet prepared for (namely, top-level-await e.g., #26626). I've been doing health checks and preparations across three's ecosystem for the last few years in preparation, and there's been a lot of velocity lately in runtimes and standardization itself. This is something I've been meaning to track in an issue, but I'm hopeful it will naturally resolve itself in parallel. We should hopefully figure out tree-shaking of WebGPU/Nodes by then, but I opted to code-split in #29404 to sidestep that matter in the meantime, although the WebGL entrypoint tree-shakes out WebGL code since #28670. I think we can agree that this is where we want to go once WebGPU is a living standard and implementations catch up.

**Comment by trusktr** [comment]
The company is not SpaceX.

And yes, the change in https://github.com/mrdoob/three.js/pull/29404 requires a build configuration update because the `three` module does not include `WebGPURenderer`. No one is trying to extort you!

**Comment by trusktr** [comment]
About top-level await, common bundlers support ESM spec nowadays, including top-level await. I agree though that if someone has a bundler or build setup that does not understand top-level await, then they will need to update their build.

Here's a question:

Can we replace top-level await with something else? For example, move it into a function. If so, then the problem will be solved in a better way.

**Comment by trusktr** [comment]
And again, because I suggested an improvement, does not mean anyone is trying to extort you @CodyJasonBennett.

The purpose of this conversation is to resolve a very real issue.

**Comment by CodyJasonBennett** [comment]
Again, you are mistaken or conflating two issues. `three` not including `WebGPURenderer` is intentional and the agreed upon resolve for the current API. Merging WebGPU code into `three` is for another discussion, one this company should really not be shy of supporting. If that means you, then this is indeed the right place to start, and perhaps we should focus into a separate issue. I maintain ecosystem support is not there for TLA, which is separate from ESM support (you can support ESM but not TLA). I can create a tracking issue, but I will not settle for mere conjecture, which long plagues these discussions, notoriously around tree-shaking. I expect tree-shaking to be a recurrent point.

**Comment by CodyJasonBennett** [comment]
Although I have personal reservations about putting experimental code into core (which is part of the absolute mess that created this issue), the technical reasons are as follows, which I explained further above in context:
- [ ] top-level-await (TLA) as WebGPU API is asynchronous for requesting a device/adapter or any sync points
- [ ] tree-shaking of Nodes which is a general open question https://github.com/mrdoob/three.js/pull/29404#discussion_r1801469444
- [x] tree-shaking of `WebGLRenderer` and helpers, which I've been routinely maintaining #28670

This list is exhaustive from a purely technical POV, but the general issue of migration and general API is ongoing, and it's a massive one. We've dealt with some of the scarier issues, like #26140, and now have `EXT_clip_control` which should remove asymmetry between backends (but still require migration). I'll leave it to the maintainers for specifics.

I will yet again maintain that it is not needed to resolve this issue, and #29404 or its outlined alternatives are sufficient. I maintain that these are two separate issues, as three API does not reflect the supposed configuration you were relying on, and the recommendation of configuration and even use in production is irresponsible and misleading. As I've seen trouble historically from unnamed companies and discourse devolve into useless conjecture, I'm willing to bet $10,000 this is and will remain the case and limit to the latest major of all build tools and runtimes (Webpack, Parcel, Rollup, Vite, ESBuild, Rspack, Turbo, SWC, Nx, Metro). As a simplification, this enables ES2021+ ESM-only output. I will leave this only until February, when I'm expecting WebGPU to become a living standard, and we can move to commit to this once implementations catch up. You will find that this is indeed very difficult, and what I'm offering is both a correction and stepping stone to a final API.

**Comment by trusktr** [comment]
> Although I have personal reservations about putting experimental code into core (which is part of the absolute mess that created this issue),

An API could be marked "experimental" in `@types/three` and in threejs.org docs. Wouldn't that be enough?

> * [ ]  top-level-await (TLA) as WebGPU API is asynchronous for requesting a device/adapter or any sync points

I believe this was fixed already. This [test repo](https://github.com/trusktr/esbuild-three-test/tree/1eb16cb2a8775937f83cfe75454f868d2c8eff3c) shows that esbuild works with target es2020, indicating no TLA is used anywhere.

> * [ ]  tree-shaking of Nodes which is a general open question [Codesplit WebGL/WebGPU entrypoints #29404 (comment)](https://github.com/mrdoob/three.js/pull/29404#discussion_r1801469444)

That's a good problem to solve. I thought tree shaking was irrespective of import/export paths. Is that not true? 

#29404 is definitely better than what we have currently because it provides a path to having both webgl and webgpu instead of a difficult split, however it still requires build tool updates downstream. 

I'm just expressing that if we can upgrade with no build changes required, it would be great: we'd just update the `three` version number, webgl APIs would be available as before, and new webgpu APIs would be available too.

**Comment by CodyJasonBennett** [comment]
I have elaborated now several times in explicit detail, and it's clear you are simply manipulating the project into pulling the trigger on merging the WebGL/WebGPU entrypoints because you, or a company that cannot be named, has configuration that incidentally relied on this being the API, after a string of irresponsible messaging that created this problem in the first place. That is not the purpose of this issue nor the current state of three API or any surrounding implementation (be it browsers, implementations, tooling—all of it). I will no longer entertain this here and expect any further progress I have enumerated to be continued elsewhere. You should have the direction you need to migrate to stable API, or continue configuration like before.

**Comment by trusktr** [comment]
@CodyJasonBennett I get the impression you don't want to discuss this, so we don't have to.

---

I made a pull request to show the concept. Now owners of this project can do what they will with this information:

- https://github.com/mrdoob/three.js/pull/29827

If it is decided that this proposal is the desired path, and more work is needed, please let me know and I will update it.

**Comment by mrdoob** [comment]
Lets focus on https://github.com/mrdoob/three.js/pull/29404
